git/gittest/repo.go

```diff
@@ -3,15 +3,84 @@ package gittest
 import (
 	"context"
 	"io/ioutil"
+	"io"
 	"os/exec"
 	"path/filepath"
 	"testing"
+	"bytes"
+	"strings"
 
 	"github.com/weaveworks/flux"
 	"github.com/weaveworks/flux/cluster/kubernetes/testfiles"
 	"github.com/weaveworks/flux/git"
 )
 
+// GPGKey creates a new, temporary GPG home directory and a public/private key
+// pair. It returns the GPG home directory, the ID of the created key, and a
+// cleanup function to be called after the caller is finished with this key.
+// Since GPG uses /dev/random, this may block while waiting for entropy to
+// become available.
+func GPGKey(t *testing.T) (string, string, func()) {
+	newDir, cleanup := testfiles.TempDir(t)
+
+	cmd := exec.Command("gpg", "--homedir", newDir, "--batch", "--gen-key")
+
+	stdin, err := cmd.StdinPipe()
+	if err != nil {
+		cleanup()
+		t.Fatal(err)
+	}
+
+	io.WriteString(stdin, "Key-Type: DSA\n")
+	io.WriteString(stdin, "Key-Length: 1024\n")
+	io.WriteString(stdin, "Key-Usage: sign\n")
+	io.WriteString(stdin, "Name-Real: Weave Flux\n")
```

Comment:
We have configuration flags for `git-user` and `git-email`, why not use them while generating the GPG key?

Reply:
Excellent point! I'll get that in.

Reply:
So we _can_, but we're only generating this key for testing that commits are get signed with the expected key ID. In actual usage of this feature, the key is provided by the user, so it's expected that it will have been generated with the desired UID. Since we don't do any key generation apart from testing, there's no reason to test that the UID matches the `git-user` and `git-email`, because we're not enforcing that anywhere.

Reply:
You are 100% right, ignore my comment.